0000-experiment-dataframe.md

```diff
@@ -0,0 +1,190 @@
+
+
+## Motivation
+With the current QE Implementation, it is very tough for novice users to tune or customize the data analysis chain because of the complexity of data structure. Data are often protected or private members, and every analysis step has own data model. Therefore it is not easy to scrutinize the actual data, and it is almost impossible to resume analysis from a particular step without hack. In addition, experimental outcome itself takes different format depending on the run options. New data structure allows users to easily access these internal data with better handling and visibility.
+
+## User Benefit
+Experimentalists will mainly enjoy the benefit of new framework. They can try fancy analysis techniques directly on the experiment data without treating it as a big deal, e.g. modifying the analysis class implementation or implementing a new class. Experiment authors (developers) will also benefit by this, especially when testing new analysis code.
+
+## Design Proposal
+
+#### Primitive data types
+
+In the conventional Qiskit execution model, we will get measure outcomes with different data types depending on run options.
+
+| Type                         | Format                 | Run option                           | Typical usage      |
+|------------------------------|------------------------|--------------------------------------|--------------------|
+| Classified, Counts (CC)      | `List[Dict[str, int]]` | `meas_level=2`                       | Standard outcome   |
+| Classified, Single shot (CS) | `List[str]`            | `meas_level=2, memory=True`          | Restless mode      |
+| Kerneled, Averaged (KA)      | `ndarray`              | `meas_level=1, meas_return="avg"`    | Device bring-up    |
+| Kerneled, Single shot (KS)   | `ndarray`              | `meas_level=1, meas_return="single"` | Discriminator cal. |
+| Raw, Averaged (RA)           | `ndarray`              | `meas_level=0, meas_return="avg"`    | Not available      |
+| Raw, Single shot (RS)        | `ndarray`              | `meas_level=0, meas_return="single"` | Not available      |
+
+The difference of these data format is one of the difficulties for developer to write custom data processor.
+Furthermore, some formats like dict and list of string are not efficient for mathematical operations.
+In QE, we should provide clean and compute-efficient data representation for these primitive data types.
+
+It is noteworthy that the bit counts can be efficiently represented by numpy array. For example, assume we received following outcome from the CC measurement.
+```python
+cc_dict = {
+  "00": 25,
+  "01": 10,
+  "10": 34,
+  "11": 13,
+}
+```
+This dictionary can be represented as follows.
+```python
+cc_array = array([25, 10, 34, 13])
+```
+Note that, for example, the index `2` is `0b10` in the binary representation, and thus we can translate bit string "10" to index 2 of the array, given we know the size of memory slot. In the same way, for the CS data,
```

Comment:
This seems unscalable to me. If I am working with 20 qubits then this implies an array with length `2**20`?

Reply:
We had the same discussion in one of past Qiskit demo days before. I agree this is not scalable, but we can use sparse array in that case because realistically we don't set shots of 2**20. Usually in QE, in the context of QCVV, we use small number of qubits so this doesn't matter for the most of cases.